corehq/apps/api/odata/serializers.py

```diff
@@ -71,3 +73,73 @@ def _clean_property_name(name):
                     properties.pop(property_name)
 
         return json.dumps(data, cls=DjangoJSONEncoder, sort_keys=True)
+
+
+class ODataXFormInstanceSerializer(Serializer):
+    """
+    A custom serializer that converts form data into an odata-compliant format.
+    Must be paired with ODataXFormInstanceResource
+    """
+    def to_json(self, data, options=None):
+        options = options or {}
+
+        domain = data.pop('domain', None)
+        if not domain:
+            raise Exception('API requires domain to be set! Did you add it in a custom create_response function?')
+
+        app_id = data.pop('app_id', None)
+        if not app_id:
+            raise Exception(
+                'API requires xmlns to be set! Did you add it in a custom create_response function?'
+            )
+
+        xmlns = data.pop('xmlns', None)
+        if not xmlns:
+            raise Exception(
+                'API requires xmlns to be set! Did you add it in a custom create_response function?'
+            )
+
+        api_path = data.pop('api_path', None)
+        if not api_path:
+            raise Exception(
+                'API requires api_path to be set! Did you add it in a custom create_response function?'
+            )
```

Comment:
I think it might be nicer if you listed all of the errors in one go. That way:

a) you move all of this validation code into its own function
b) the user can, in one fell swoop, edit and send back a well formed response

Reply:
So the thing about these exceptions (copied from ```ODataCommCareCaseSerializer```) is that they will fail due to programmer error rather than user error.  Probably assertions are the more appropriate thing to use here.

Reply:
3842f4865fb